### PR TITLE
[compiler-rt] Don't rely on automatic codesigning with Apple's linker

### PR DESCRIPTION
**Pull request by bdash**
In https://github.com/llvm/llvm-project/pull/88323, I changed the logic within `add_compiler_rt_runtime` to only explicitly code sign the resulting library if an older version of Apple's ld64 was in use. This was based on the assumption that newer versions of ld64 and the new Apple linker always ad-hoc sign their output binaries. This is true in most cases, but not when using Apple's new linker with the `-darwin-target-variant` flag to build Mac binaries that are compatible with Catalyst.

Rather than adding increasingly complicated logic to detect the exact scenarios that require explicit code signing, I've opted to always explicitly code sign when using any Apple linker. We instead detect and use the 'linker-signed' codesigning option when possible to match the signatures that the linker would otherwise create. This avoids having non-'linker-signed' ad-hoc signatures which was the underlying problem that https://github.com/llvm/llvm-project/pull/88323 was intended to address.

### CODE REVIEW

**Comment by bdash** [comment]
@cjappl, this should address the issue you mentioned in https://github.com/llvm/llvm-project/pull/88323.

**Comment by bdash** [comment]
@yln, would you mind reviewing this follow-up to https://github.com/llvm/llvm-project/pull/88323?

**Comment by cjappl** [comment]
I can confirm that this fixes the problem I reported. I am now able to build and run `check-asan` and others with no code-signing issues on the two macs I have access to.

Much appreciated @bdash

File: compiler-rt/cmake/Modules/AddCompilerRT.cmake [hunk 1]
```diff
@@ -387,35 +387,35 @@ function(add_compiler_rt_runtime name type)
...
       if (APPLE AND NOT CMAKE_LINKER MATCHES ".*lld.*")
-        # Ad-hoc sign the dylibs when using Xcode versions older than 12.
-        # Xcode 12 shipped with ld64-609.
-        # FIXME: Remove whole conditional block once everything uses Xcode 12+.
-        set(LD_V_OUTPUT)
+        # Apple's linker signs the resulting dylib with an ad-hoc code signature in
+        # most situations, except:
+        # 1. Versions of ld64 prior to ld64-609 in Xcode 12 predate this behavior.
+        # 2. Apple's new linker does not when building with `-darwin-target-variant`
+        #    to support macOS Catalyst.
```
Comment (yln) [inline compiler-rt/cmake/Modules/AddCompilerRT.cmake:394]: According to this (and assuming we can drop support for the old linker) would be to ensure the linker behaves consistently and always codesigns, right?

(Not asking for this in this PR, just wanted to make sure I understand.)

**Review by yln — APPROVED** [review]

**Comment by yln** [comment]
LGTM, thanks @bdash!

File: compiler-rt/cmake/Modules/AddCompilerRT.cmake [hunk 2]
```diff
@@ -387,35 +387,35 @@ function(add_compiler_rt_runtime name type)
...
+        # behaves as expected when processed by subsequent tooling.
+        #
+        # Detect whether `codesign` supports `-o linker-signed` by passing it as an
+        # argument and looking for `invalid argument "linker-signed"` in its output.
+        # FIXME: Remove this once all supported toolchains support `-o linker-signed`.
         execute_process(
-          COMMAND sh -c "${CMAKE_LINKER} -v 2>&1 | head -1"
-          RESULT_VARIABLE HAD_ERROR
-          OUTPUT_VARIABLE LD_V_OUTPUT
+          COMMAND sh -c "codesign -f -s - -o linker-signed this-does-not-exist 2>&1 | grep -q linker-signed"
```
[the remark this answers is not in the document]
Comment (bdash) [reply compiler-rt/cmake/Modules/AddCompilerRT.cmake:404]: I wish there were a more direct way to test for support, but this was the best I could come up with. I've verified the format of errors for unknown `-o` values is the same from `codesign` on macOS 10.15 through macOS 14.4, so hopefully it'll stay consistent going forwards too.

**Comment by bdash** [comment]
@yln What needs to be done to merge this? Does someone with commit access need to hit the merge button?

**Comment by yln** [comment]
Ah, I didn't realize to you don't have PR merge privileges / commit access.

I will merge this for you on Tue morning (Mo is holiday in US) to avoid landing this just before the weekend reduce friction.  Feel free to ping me if I haven't done this by Tue evening.

If something goes wrong, people will comment here (and are allowed to revert our change).  Please respond in a timely matter.

Adding reviewers for visibility: @thetruestblue @wrotki

**Comment by bdash** [comment]
Sounds good, thank you!

**Comment by bdash** [comment]
@yln Any chance you'll be able to get it in this week?

**Pull request merged**